sdk/opendp/smartnoise/_ast/expression.py

```diff
@@ -70,9 +89,9 @@ def column_name(self):
             parts = self.expression.name.split(".")
             return parts[0] if len(parts) == 1 else parts[1]
         elif type(self.expression) is AllColumns:
-            return "???"
+            return Identifier("???")
```

Comment:
This line will fix the AST type of the `column_names` return value for `AllColumns` be consistent with the return type when called on any `NamedExpression` which has an alias name specified.  Both will return `Identifier` after this fix.  However, I believe we will still return `str` for unaliased columns (line 90 of this file), and for unaliased aggregates (using `symbol_name`).

I believe this is only used in the code as the key in the tuple of (name, symbol) under `m_symbols` hash table.  The code seems to be working OK currently, despite the hash table keys being a mix of `str` and `Identifier`, but this potentially makes us vulnerable to bugs.

This raises the question of what we want the return type of `column_name` to be.  The `column_name` defines the output column name which will be used when serializing the result rowset, and isn't part of the query AST.  When generating the output rowset, column names are always converted to `str` before use.  However, ensuring that all `column_name` outputs are `Identifier` could allow us to do some runtime validation to ensure that column names always adhere to the grammar definition of `IDENT`.

I don't have a strong preference either way.  Thoughts?

A second issue is that this particular line is underspecified, because `column_name` doesn't really have a meaning for `AllColumns`.  The output column names for any `SELECT *` will always expand to the full list of columns in the database, so it should be impossible to ever have a column name that is `???`.  During symbol resolution, `AllColumns` are expanded (metadata are required for the parser to know what column names comprise `*`), the symbol resolution will resolve the symbol as a single column of type `AllColumns`.  So we theoretically don't need this line.  It could serve as a failsafe if anyone tries to use a parsed query without symbols, but I'm wondering if we should just throw an exception in that case anyway?  I can't think of a case where calling `column_name` on `AllColumns` would not be an error.

Reply:
You are right
I think this would be better to keep it like that for the moment and have latter one PR dedicated to avoiding  mixing `str` and `Identifier` because the changes would be trickier (for example to avoid to call `Identifier` on another `Identifier` in the Scope.push_name method).

Reply:
I just realized, we will never have the issue about `AllColumns` I raised above, because we don't support `SELECT *`.  And we do need to return `symbol_name` on `AllColumns`, because it's used for example in `SELECT COUNT(*)`, which resolves to `count_<AllColumns.symbol_name>`